utils/gen-snapshot-tests.js

```diff
@@ -95,41 +96,100 @@ function generateTestFile(root, examples, componentFileMap) {
+    if (!alreadyMapped) {
+        // Only gather examples for this component file if
+        // we didn't see it already.
+        collation.componentFileMap[match[1]] = file;
+
+        const componentDoc = path.join(
+            path.dirname(file),
+            path.basename(file, path.extname(file)),
+        ) + ".md";
+
+        collation.examples = collation.examples.concat(
+            readExamplesFromDocument(componentDoc),
+        );
+    }
+}
+
+function extractExamplesAndComponentFiles(section, collation) {
+    // Add any examples fromt he section content.
+    collation.examples = collation.examples.concat(
+        readExamplesFromDocument(section.content),
+    );
+
+    // Then process any component files and add their stuff too.
+    if (section.components) {
+        const componentGlobs = typeof section.components === "string"
+            ? [section.components]
+            : section.components;
+        for (const componentGlob of componentGlobs) {
+            const files = glob.sync(componentGlob, {ignore: ignored});
+            for (const file of files) {
+                extractExamplesAndComponentsForFile(file, collation);
+            }
+        }
     }
 
-    if (section.content) {
-        const root = path.dirname(section.content);
-
-        if (section.components) {
-            glob(section.components, {ignore: ignored}, (err, files) => {
-                const componentFileMap = {};
-                for (const file of files) {
-                    const src = fs.readFileSync(file, "utf8");
-                    const match =
-                        src.match(/export default class \s*(\w+)/) ||
-                        src.match(/export default \s*\w+\(\s*(\w+)/);
-                    componentFileMap[match[1]] = file;
-                }
-
-                generateTestFile(root, examples, componentFileMap);
-            });
-        } else {
-            generateTestFile(root, examples);
+    if (section.sections) {
+        // Repeat all this for each sub-section.
+        for (const childSection of section.sections) {
+            extractExamplesAndComponentFiles(
+                childSection,
+                collation,
+            );
         }
-    } else {
+    }
+}
+
+function tryGenerateSectionTests(section) {
+    // If there is no content nor sections, then we should quit.
+    if (!section.content && !section.sections) {
         // eslint-disable-next-line no-console
-        console.warn("no content for section: ");
+        console.warn("no content or sub-sections for section: ");
         // eslint-disable-next-line no-console
         console.warn(JSON.stringify(section));
+        return;
+    }
+
+    // Now, let's collate all our examples and the components for the section
+    // they're in.
+    const collation = {
+        examples: [],
+        componentFileMap: {},
+    };
+    extractExamplesAndComponentFiles(section, collation);
```

Comment:
Could `extractExamplesAndComponentFiles` return a `collation` instead of mutating the arguments.

Reply:
I had played around with that. We could easily change to that model, i think. I'll take a look.